File: scripts/generate-rdf.py
from pathlib import Path
from hashlib import sha256
from importlib.metadata import version as package_version
import json

from pyld import jsonld
from rdflib import Graph

root = Path(__file__).resolve().parents[1]
source = root / 'src/data/semantic/knowledge-graph.jsonld'
target = root / 'src/data/semantic/knowledge-graph.ttl'
lock_path = root / 'src/data/semantic/rdf-lock.json'
release = json.loads((root / 'src/data/release.json').read_text(encoding='utf-8'))

g = Graph()
g.parse(source, format='json-ld')

# Serialize the RDF graph to an arbitrary blank-node-labelled N-Triples stream,
# then apply a dataset canonicalization algorithm that deterministically assigns
# blank-node identifiers.  Sorting canonical lines is an additional stable byte
# ordering guard.  Because this project uses the default graph only, canonical
# N-Quads output is also valid N-Triples and therefore valid Turtle syntax.
raw_ntriples = g.serialize(format='nt')
canonical_nquads = jsonld.normalize(
    raw_ntriples,
    {
        'algorithm': 'URDNA2015',
        'inputFormat': 'application/n-quads',
        'format': 'application/n-quads',
    },
)
lines = sorted(line for line in canonical_nquads.splitlines() if line.strip())
text = '\n'.join(lines) + '\n'
target.write_text(text, encoding='utf-8')

# Parse the emitted distribution again.  A broken canonicalizer/output or an
# accidental named-graph introduction must never ship as the .ttl distribution.
check = Graph()
check.parse(target, format='nt')
if len(check) != len(g):
    raise SystemExit(f'RDF round-trip mismatch: source={len(g)} ttl={len(check)}')

lock = {
    'release': release['release'],
    'source': 'src/data/semantic/knowledge-graph.jsonld',
    'distribution': 'src/data/semantic/knowledge-graph.ttl',
    'triples': len(g),
    'jsonldSha256': sha256(source.read_bytes()).hexdigest(),
    'ttlSha256': sha256(target.read_bytes()).hexdigest(),
    'canonicalizedBlankNodes': True,
    'canonicalizationAlgorithm': 'URDNA2015',
    'canonicalizationLibrary': 'PyLD',
    'canonicalizationLibraryVersion': package_version('PyLD'),
    'canonicalSyntax': 'canonical-default-graph-nquads-as-ntriples-compatible-turtle',
    'generated': release['dateModified'],
}
lock_path.write_text(json.dumps(lock, indent=2) + '\n', encoding='utf-8')
print(json.dumps(lock, indent=2))

File: scripts/generate-rdf.mjs
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import {generatedWorkspace} from './generated-workspace.mjs';
import {canonicalizeRdfDocument,RDF_MEASUREMENT_CONTRACT} from './lib/rdf-measurement.mjs';

const source='src/data/semantic/knowledge-graph.jsonld';
const generated=generatedWorkspace();
const target=path.join(generated.semantic,'knowledge-graph.ttl');
const lockPath=path.join(generated.semantic,'rdf-lock.json');
fs.mkdirSync(generated.semantic,{recursive:true});
const release=JSON.parse(fs.readFileSync('src/data/release.json','utf8'));
const doc=JSON.parse(fs.readFileSync(source,'utf8'));
const measurement=await canonicalizeRdfDocument(doc);
fs.writeFileSync(target,measurement.text);
const sha=b=>crypto.createHash('sha256').update(b).digest('hex');
const distribution='.generated/semantic/knowledge-graph.ttl';
const lock={
  release:release.release,
  source,
  distribution,
  triples:measurement.triples,
  jsonldSha256:sha(fs.readFileSync(source)),
  ttlSha256:sha(fs.readFileSync(target)),
  canonicalizedBlankNodes:true,
  ...RDF_MEASUREMENT_CONTRACT,
  generated:release.dateModified,
};
fs.writeFileSync(lockPath,JSON.stringify(lock,null,2)+'\n');
console.log(JSON.stringify(lock,null,2));
